Objects/perf_trampoline.c

```diff
@@ -0,0 +1,233 @@
+    size_t code_size;
+    struct code_arena_st *prev;
+};
+
+typedef struct code_arena_st code_arena_t;
+
+static Py_ssize_t extra_code_index = -1;
+static code_arena_t *code_arena;
+static FILE *perf_map_file;
+
+static int
+new_code_arena(void)
+{
+    // non-trivial programs typically need 64 to 256 kiB.
+    size_t mem_size = 4096 * 16;
+    assert(mem_size % sysconf(_SC_PAGESIZE) == 0);
+    char *memory = mmap(NULL,  // address
+                        mem_size,
+                        PROT_READ | PROT_WRITE,
+                        MAP_PRIVATE | MAP_ANONYMOUS,
+                        -1,  // fd (not used here)
+                        0);  // offset (not used here)
+    if (!memory) {
+        Py_FatalError("Failed to allocate new code arena");
+        return -1;
+    }
+    void *start = &_Py_trampoline_func_start;
+    void *end = &_Py_trampoline_func_end;
+    size_t code_size = end - start;
+
+    size_t n_copies = mem_size / code_size;
+    for (size_t i = 0; i < n_copies; i++) {
+        memcpy(memory + i * code_size, start, code_size * sizeof(char));
+    }
+
+    mprotect(memory, mem_size, PROT_READ | PROT_EXEC);
+
+    code_arena_t *new_arena = PyMem_RawCalloc(1, sizeof(code_arena_t));
+    if (new_arena == NULL) {
+        Py_FatalError("Failed to allocate new code arena struct");
+        return -1;
+    }
+
+    new_arena->start_addr = memory;
+    new_arena->current_addr = memory;
+    new_arena->size = mem_size;
+    new_arena->size_left = mem_size;
+    new_arena->code_size = code_size;
+    new_arena->prev = code_arena;
+    code_arena = new_arena;
+    return 0;
+}
+
+static void
+free_code_arenas(void)
+{
+    code_arena_t *cur = code_arena;
+    code_arena_t *prev;
+    code_arena = NULL; // invalid static pointer
+    while(cur) {
+        munmap(cur->start_addr, cur->size);
+        prev = cur->prev;
+        PyMem_RawFree(cur);
+        cur = prev;
+    }
+}
+
+static inline py_trampoline
+code_arena_new_code(code_arena_t *code_arena)
+{
+    py_trampoline trampoline = (py_trampoline)code_arena->current_addr;
+    code_arena->size_left -= code_arena->code_size;
+    code_arena->current_addr += code_arena->code_size;
+    return trampoline;
+}
+
+static inline py_trampoline
+compile_trampoline(void)
+{
+    if ((code_arena == NULL) || (code_arena->size_left <= code_arena->code_size)) {
+        if (new_code_arena() < 0) {
+            return NULL;
+        }
+    }
+
+    assert(code_arena->size_left <= code_arena->size);
+    return code_arena_new_code(code_arena);
+}
+
+static inline FILE *
+perf_map_get_file(void)
+{
+    if (perf_map_file) {
+        return perf_map_file;
+    }
+    char filename[100];
+    pid_t pid = getpid();
+    // TODO: %d is incorrect if pid_t is long long
+    snprintf(filename, sizeof(filename), "/tmp/perf-%d.map", pid);
+    perf_map_file = fopen(filename, "a");
+    if (!perf_map_file) {
+        _Py_FatalErrorFormat(__func__, "Couldn't open %s: errno(%d)", filename, errno);
```

Comment:
If map file opening calls fail it is probably best to just disable the profiling rather than reporting the error and crashing. That isn't so important for the basic -Xperf dedicated profiled process case, but for turning it on and off at runtime you never want turning it on to crash the program. A `-v` verbose warning with error details instead of the fatal error would work to help people debug if they're surprised about no map existing.

Reply:
It also avoids DoS scenarios where something just creates a bunch of /tmp/perf-XXX.map files and blocks the start of any python process with perf profiling enabled.

Reply:
We can maybe throw an unraisable exception instead. I still think we should give more visibility to this error even when -v is not used.